### ISSUE TITLE
Custom Keras model adds extra dimension to scalar inputs

### ISSUE THREAD

**Issue by keyonvafa** [tensorflow/tensorflow#39774]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): Yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Ubuntu 16.04.6
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device: N/A
- TensorFlow installed from (source or binary): Binary
- TensorFlow version (use command below): 2.2.0
- Python version: 3.7.3
- Bazel version (if compiling from source):
- GCC/Compiler version (if compiling from source):
- CUDA/cuDNN version: 10.1
- GPU model and memory: TITAN V (12036MiB)

**Describe the current behavior**

I am creating a custom Keras model that takes scalar inputs (i.e. inputs have shape [batch_size]). When I compile the model and pass in a tf.data.Dataset to model.fit(), the shape of the input data is changed to [batch_size, 1]. 

However, when I train in a training loop using an iterator, the shape of the input data is [batch_size]. 

This happens both locally and on Colab.

**Describe the expected behavior**
The current behavior adds a side effect of changing the data input shape when using model.fit(). This behavior is inconsistent, since an extra dimension is not added to scalar inputs when training iteratively. The expected behavior would not add an extra dimension to scalar inputs, or at least be consistent between model.fit() and iterative training.

**Standalone code to reproduce the issue**
Colab Notebook: https://colab.research.google.com/drive/14RyEYCRtYCwryJhkd4koYnK04RlUoqrI?usp=sharing

```python
import numpy as np
import tensorflow as tf

class ToyModel(tf.keras.Model):
  def call(self, inputs):
    # Inputs are scalar, so we should only have a batch dimesion.
    print("Input shape is {}".format(inputs.shape))
    return tf.reduce_sum(inputs)

toy_model = ToyModel()
optimizer = tf.keras.optimizers.Adam(learning_rate=1e-3)
loss = tf.keras.losses.MeanSquaredError()

data_size = 10
data_inputs = np.float32(np.random.randn(data_size))
data_outputs = np.float32(np.random.randn(data_size))
dataset = tf.data.Dataset.from_tensor_slices(
    (data_inputs, data_outputs)).batch(data_size)

# Train iteratively
iterator = iter(dataset)
inputs, outputs = iterator.next()
with tf.GradientTape() as tape:
  predictions = toy_model(inputs)
  loss_value = loss(outputs, predictions)
grads = tape.gradient(loss_value, toy_model.trainable_weights)
optimizer.apply_gradients(zip(grads, toy_model.trainable_weights))
# Input shape is [batch_size]

# Now, try training with fit() instead
toy_model.compile(optimizer=optimizer, loss=loss)
toy_model.fit(dataset, epochs=1)  
# Input shape is [batch_size, 1]

```

**Comment by amahendrakar** [comment]
Was able to reproduce the issue with [TF v2.2](https://colab.research.google.com/gist/amahendrakar/149b5e8291940bc45ef9491e7d7f9454/39774.ipynb) and [TF-nightly](https://colab.research.google.com/gist/amahendrakar/a82c862bd1db9fbbe3045c7e57cb68fc/39774-tf-nightly.ipynb). Please find the attached gist. Thanks!

**Comment by jvishnuvardhan** [comment]
@keyonvafa I think this is not a bug but intended behavior. Under the hood, some functionality of `model.fit` require the shape to be like that but it will not change your original data shape. Does this concern you for your use-case? Can you please mention little more details how this affects your use-case? Thanks!

**Comment by keyonvafa** [comment]
@jvishnuvardhan Thanks for the response. The main issue for me is the inconsistency. Sometimes when I'm prototyping I'll train iteratively so I can debug in Eager mode. When I run it in the terminal I'll change training to `model.fit`, but I also have to include lines like `inputs = inputs[:, 0]` and `outputs = outputs[:, 0]`.

**Comment by jvishnuvardhan** [comment]
@keyonvafa Is this still an issue for you? Are you interested in contributing to update any part of the code. Any update need to be backward compatible. Thanks!

**Comment by keyonvafa** [comment]
@jvishnuvardhan Thanks for the note. For now I'm avoiding this issue by only using iterative training. I unfortunately don't have the bandwidth to contribute to the code

**Comment by jvishnuvardhan** [comment]
@keyonvafa Do you want to close the issue or keep it open? Thanks!

**Comment by keyonvafa** [comment]
@jvishnuvardhan It seems like others are having the same issue so I'm fine keeping it open. Thanks!

**Comment by jvishnuvardhan** [comment]
Sure. I am adding `contributions welcome` labels so that community can pick it up and contribute. Thanks!

**Comment by tensorflowbutler** [comment]
Hi There,

 We are checking to see if you still need help on this issue, as you are using an older version of tensorflow(1.x) which is officially considered as end of life. We recommend that you upgrade to 2.4 or later version and let us know if the issue still persists in newer versions. 

 This issue will be closed automatically 7 days from now. If you still need help with this issue, Please open a new issue for any help you need against 2.x, and we will get you the right help.

**Comment by keyonvafa** [comment]
I've been having this issue with Tensorflow 2.2, not Tensorflow 1.x.

**Comment by sachinprasadhs** [comment]
@keyonvafa , That was wrongly triggered message, please ignore it. Thanks!

**Comment by sushreebarsa** [comment]
Was able to replicate the issue with TF 2.5,please find the gist [here](https://colab.research.google.com/gist/sushreebarsa/be51a75a08d8426a19e4c05a49119035/untitled295.ipynb) ..Thanks!

**Comment by blackyang** [comment]
> I think this is not a bug but intended behavior.
@jvishnuvardhan i think this is clearly a bug right? because it's inconsistent behavior between using custom Model vs using functional API. Could you fix this? Thx.

**Comment by tensorflowbutler** [comment]
Hi There,

 This is a stale issue. As you are using an older version of tensorflow, we are checking to see if you still need help on this issue. Please test the issue with the latest TensorFlow (TF2.7 and tf-nightly). If the issue still persists with the newer versions of TF, please feel free to open it in [keras-team/keras](https://github.com/keras-team/keras/issues) repository by providing details about the issue and a standalone code to reproduce the issue. Thanks! 

 Please note that Keras development has moved to a separate Keras-team/keras repository to focus entirely on only Keras. Thanks!
